### Monty Hall Problem
Suppose you're on a game show, and you're given the choice of three doors: Behind one door is a car; behind the others, goats. You pick a door, say No. 1, and the host, who knows what's behind the doors, opens another door, say No. 3, which has a goat. He then says to you, **"Do you want to pick door No. 2?" Is it to your advantage to switch your choice?**

*The problem was originally posed (and solved) in a letter by **Steve Selvin** to the American Statistician in 1975. It became famous as a question from reader Craig F. Whitaker's letter quoted in **Marilyn vos Savant's** "Ask Marilyn" column in Parade magazine in 1990* -Wikipedia

### Bayesian Network
A Bayesian network is a probabilistic graphical model that represents a set of variables and their **conditional dependencies** via a **directed acyclic graph (DAG)**. Bayesian networks are ideal for taking an event that occurred and predicting the likelihood that any one of several possible known causes was the contributing factor.

### Solution

Create a Directed Acyclic Graph with the following nodes:
- Guest Door: Door chosen by guest
- Prize: Door containing the prize
- Monty: Door chosen by Monty

In [1]:
'''
Three doors: A,B,C

--------------       --------------
| Guest door |       | Prize door |
--------------       --------------
        \                  /
         \                /
          \              /
           \            /
           --------------
           | Monty door |
           --------------
            
'''
print("Directed Acyclic Graph")

Directed Acyclic Graph


In [2]:
# !pip install pomegranate==v0.14.8
# # !pip install pomegranate==0.6.0
# #!pip install pomegranate

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.3/4.3 MB 14.6 MB/s eta 0:00:0000:010:01m
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
  Created wheel for pomegranate: filename=pomegranate-0.14.8-cp310-cp310-linux_x86_64.whl size=8998179 sha256=66521460eb8b6f0437e124e7d7cc58e66b2188c929d46b94eb3e87ae519ccbde
  Stored in directory: /root/.cache/pip/wheels/72/a0/96/f35659c8c85a4b0b2be7333a70048999109ce1a39d1064118e
Successfully built pomegranate


In [10]:
# import math
# import pomegranate
# import pomegranate.distributions 
# from pomegranate import BayesianNetwork



# guest= DiscreteDistribution({'A': 1./3,'B': 1./3,'C': 1./3})
# prize= DiscreteDistribution({'A': 1./3,'B': 1./3,'C': 1./3})

In [11]:
# monty= ConditionalProbabilityTable(
#     [['A', 'A', 'A', 0.0],
#      ['A', 'A', 'B', 0.5],
#      ['A', 'A', 'C', 0.5],
#      ['A', 'B', 'A', 0.0],
#      ['A', 'B', 'B', 0.0],
#      ['A', 'B', 'C', 1.0],
#      ['A', 'C', 'A', 0.0],
#      ['A', 'C', 'B', 1.0],
#      ['A', 'C', 'C', 0.0],
#      ['B', 'A', 'A', 0.0],
#      ['B', 'A', 'B', 0.0],
#      ['B', 'A', 'C', 1.0],
#      ['B', 'B', 'A', 0.5],
#      ['B', 'B', 'B', 0.0],
#      ['B', 'B', 'C', 0.5],
#      ['B', 'C', 'A', 0.0],
#      ['B', 'C', 'B', 0.0],
#      ['B', 'C', 'C', 0.0],
#      ['C', 'A', 'A', 0.0],
#      ['C', 'A', 'B', 1.0],
#      ['C', 'A', 'C', 0.0],
#      ['C', 'B', 'A', 1.0],
#      ['C', 'B', 'B', 0.0],
#      ['C', 'B', 'C', 0.0],
#      ['C', 'C', 'A', 0.5],
#      ['C', 'C', 'B', 0.5],
#      ['C', 'C', 'C', 0.0]],
#      [guest, prize]
# )

In [12]:
# network= BayesianNetwork("Monty Hall Problem")

In [22]:
# n1= pomegranate.Node(guest, "guest")
# n2= pomegranate.Node(prize, "prize")
# n3= pomegranate.Node(monty, "monty")

In [23]:
# network.add_states(n1, n2, n3)
# network.add_edge(n1, n3)
# network.add_edge(n2, n3)
# network.bake()

In [24]:
# beliefs= network.predict({'guest': 'A'})
# beliefs= map(str, beliefs)
# print("\n".join("{}\t{}".format(state.name, belief) for state, belief in zip(network.states, beliefs)))

AttributeError: 'NoneType' object has no attribute 'log_probability'

In [ ]:
# beliefs= network.predict_proba({'guest':'A', 'monty':'B'})
# beliefs= map(str, beliefs)
# print("\n".join("{}\t{}".format(state.name, belief) for state, belief in zip(network.states, beliefs)))

In [30]:
!pip install pomegranate

In [39]:
import math
import pomegranate 
import pomegranate as pg

# The door chosen by the guest can be any of the three
# The door which has the prize behind it, can be any of the three
guest= pg.DiscreteDistribution({'A':1./3,'B':1./3,'C':1./3})
prize= pg.DiscreteDistribution({'A':1./3,'B':1./3,'C':1./3})

monty=pg.ConditionalProbabilityTable(
   [['A','A','A',0.0],
    ['A','A','B',0.5],
    ['A','A','C',0.5],
    ['A','B','A',0.0],
    ['A','B','B',0.0],
    ['A','B','C',1.0],
    ['A','C','A',0.0],
    ['A','C','B',1.0],
    ['A','C','C',0.0],
    ['B','A','A',0.0],
    ['B','A','B',0.0],
    ['B','A','C',1.0],
    ['B','B','A',0.5],
    ['B','B','B',0.0],
    ['B','B','C',0.5],
    ['B','C','A',1.0],
    ['B','C','B',0.0],
    ['B','C','C',0.0],
    ['C','A','A',0.0],
    ['C','A','B',1.0],
    ['C','A','C',0.0],
    ['C','B','A',1.0],
    ['C','B','B',0.0],
    ['C','B','C',0.0],
    ['C','C','A',0.5],
    ['C','C','B',0.5],
    ['C','C','C',0.0]
   ],
    [guest,prize])

n1= pg.Node(guest,name="guest")
n2= pg.Node(prize,name="prize")
n3= pg.Node(monty,name="monty")

model= pg.BayesianNetwork("Monty Hall Problem")

model.add_states(n1,n2,n3)
model.add_edge(n1,n3)
model.add_edge(n2,n3)

model.bake()

#beliefs=model.predict({'guest':'A'})
beliefs= model.predict_proba([['A', None, None]]) #[guest,prize,monty]
beliefs= map(str,beliefs)
print("\n".join("{}\t{}".format(state.name,belief)for state,belief in zip(model.states,beliefs)))

print("]n------------------------------------------------------------------\n")

beliefs= model.predict_proba([['A',None,'B']]) #[guest,prize,monty]
beliefs= map(str,beliefs)
print("\n".join("{}\t{}".format(state.name,belief)for state,belief in zip(model.states,beliefs)))

guest	['A' {
         "class" : "Distribution",
         "dtype" : "str",
         "name" : "DiscreteDistribution",
         "parameters" : [
             {
                 "A" : 0.3333333333333333,
                 "B" : 0.3333333333333333,
                 "C" : 0.3333333333333333
             }
         ],
         "frozen" : false
     }
 {
     "class" : "Distribution",
     "dtype" : "str",
     "name" : "DiscreteDistribution",
     "parameters" : [
         {
             "A" : 0.0,
             "C" : 0.49999999999999983,
             "B" : 0.49999999999999983
         }
     ],
     "frozen" : false
 }                                     ]
]n------------------------------------------------------------------

guest	['A' {
         "class" : "Distribution",
         "dtype" : "str",
         "name" : "DiscreteDistribution",
         "parameters" : [
             {
                 "A" : 0.3333333333333334,
                 "B" : 0.0,
                 "C" : 0.6666666666666664
   

#### Case 1: ['A', None, None] 
Guest has picked Door A. Monty has not picked any door<br>
so the prize may be in any of the 3 doors(A or B or C): 
<br>
**"A" : 0.3333333333333333<br>
"B" : 0.3333333333333333<br>
"C" : 0.3333333333333333**
<br>
<br>
Since Guest has picked Door A, Monty cannot choose Door A, he can only choose either B or C<br>
so monty may choose any of the 2 remaining doors (B or C): 
<br>
**"A" : 0.0<br>
"C" : 0.49999999999999983<br>
"B" : 0.49999999999999983**
<br>
<br>
#### Case 2: ['A', None, 'B']
Guest has picked Door A. Monty has picked Door B<br>
Results:<br>
**"A" : 0.3333333333333334<br>
"B" : 0.0<br>
"C" : 0.6666666666666664**
<br><br>
The probability that the prize is behind door A is 0.333 and behind door C is 0.666.
<br>
*Savant's response was that the guest should switch to the other door. Under the standard assumptions, <br>the switching strategy has a 
2/3 probability of winning the prize, while the strategy of sticking with the initial choice has only a 1/3 probability.*


### For the Guest to maximize his probability of winning the prize, He should switch his choice from Door A to Door C.
